src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicallyResolvedFunctionCallIterator.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Authors: Stefan Irimescu, Can Berker Cikis
+ *
+ */
+
+package sparksoniq.jsoniq.runtime.iterator.functions;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.rumbledb.api.Item;
+import sparksoniq.exceptions.IteratorFlowException;
+import sparksoniq.jsoniq.item.FunctionItem;
+import sparksoniq.jsoniq.runtime.iterator.HybridRuntimeIterator;
+import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
+import sparksoniq.jsoniq.runtime.iterator.functions.base.FunctionIdentifier;
+import sparksoniq.jsoniq.runtime.iterator.functions.base.Functions;
+import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+import sparksoniq.semantics.DynamicContext;
+
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class DynamicallyResolvedFunctionCallIterator extends HybridRuntimeIterator {
+
+    private static final long serialVersionUID = 1L;
+    // parametrized fields
+    private FunctionIdentifier _functionIdentifier;
+    private List<RuntimeIterator> _functionArguments;
+
+    // calculated fields
+    private FunctionItem _functionItem;
+    private RuntimeIterator _functionCallIterator;
+    private Item _nextResult;
+
+
+    public DynamicallyResolvedFunctionCallIterator(
+            FunctionIdentifier functionIdentifier,
+            List<RuntimeIterator> functionArguments,
+            IteratorMetadata iteratorMetadata
+    ) {
+        super(null, iteratorMetadata);
+        _functionIdentifier = functionIdentifier;
+        _functionArguments = functionArguments;
+
+    }
+
+    @Override
+    public void openLocal() {
+        _functionCallIterator = Functions.getUserDefinedFunctionIterator(
+            _functionIdentifier,
+            getMetadata(),
+            _functionArguments
+        );
```

Comment:
If I'm not mistaken the assignment to _functionCalliterator is redundant. This lookup is already performed in initIsRdd (which is always called before openLocal() or getRDDAux() ) method and saved to class state.

Reply:
Good catch, thanks.

Note to self: resolving the function at init time may prevent recursive call. We may need to move it back to open and getRDD later on to enable recursion.